frontend/lib/resolution/scope-queries.cpp

```diff
@@ -701,8 +702,24 @@ static void errorIfNameNotInScope(Context* context,
                              checkedScopes, result);
 
   if (got == false || result.size() == 0) {
-    CHPL_REPORT(context, UseImportUnknownSym, clauseForError, scope, useOrImport,
-                isRename, name.c_str());
+    CHPL_REPORT(context, UseImportUnknownSym, exprForError, clauseForError,
+                scope,
+                useOrImport,
+                name.c_str(),
+                isRename);
+    return;
+  }
+
+  // If there is a single ID, then go ahead and try to emit warnings. If
+  // not, it's an overloaded routine, or it will get an ambiguity error.
+  if (result.size() == 1 && result[0].numIds() == 1) {
+    ID idTarget = result[0].firstId();
+    if (!idTarget.isEmpty()) {
+      ID idMention = exprForError->id();
+      CHPL_ASSERT(!idMention.isEmpty());
+      parsing::deprecationWarningForId(context, idMention, idTarget);
```

Comment:
Since you call this whether or not there is a deprecation warning for idTarget, it's going to create an entry in a table (for the query). This makes me think it would probably be better to either separately check for an Attribute or to make these not be queries. So, you might consider making `deprecationWarningForId` not be a query itself, but something that runs `idToAttributes`. It can run an implementation-helper query after that, if you need it to.

Reply:
If you make it not a query itself, you can also give it the responsibility for raising the error, to make it easy to have a version that returns the error if that is needed in the future.

Reply:
These things are already wrappers around queries that try their best not to commit to the cache if possible.